Implement ALTER DATABASE {{ADD,DROP,MODIFY} TABLE OVERRIDE

stigsb opened the pull request:
Changes to non-storage overrides are made effective by running ALTER TABLE reconciliation queries. (Storage overrides affect fundamental table storage, such as ORDER BY, while for example adding a skipping index is not considered a storage modification.) Modifications that change storage parameters require dropping the database and re-syncing, doing this must be enabled explicitly with the `allow_alter_database_to_drop_and_recreate_if_needed=1` setting.

Also implemented ALTER DATABASE RESET SETTING.

Refactored database alters to use a separate AlterDatabaseCommands class with individual classes for handling different commands.

Included @larspars 's ReplicationHelper class for MaterializedMySQL integration tests.

Cleaned up MaterializedMySQL integration tests a bit, now runs only one ClickHouse instance instead of three.


### Changelog category (leave one):
- Improvement

### Changelog entry
 - Add ALTER DATABASE {ADD,DROP,MODIFY} TABLE OVERRIDE (currently only supported for MaterializedMySQL)
 - Add ALTER DATABASE RESET SETTING (currently only supported for MaterializedMySQL)

### Documentation entry for user-facing changes
- [ ] Documentation is written

larspars committed:
Add ReplicationHelper class for MaterializedMySQL integration tests

stigsb committed:
Implement ALTER DATABASE {{ADD,DROP,MODIFY} TABLE OVERRIDE,RESET SETTING}

Changes to non-storage overrides are made effective by running ALTER TABLE
reconciliation queries. (Storage overrides affect table storage parameters
such as ORDER BY.) Modifications that change storage parameters require
dropping the database and re-syncing, doing this must be enabled explicitly
with the `allow_alter_database_to_drop_and_recreate_if_needed=1` setting.

Implemented ALTER DATABASE RESET SETTING.

Refactored database alters to use a separate AlterDatabaseCommands
class with individual classes for handling different commands.

Added a ReplicationHelper class for MaterializedMySQL integration tests.

Cleaned up MaterializedMySQL integration tests a bit, now runs only one
ClickHouse instance instead of three.

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #54219 from stigsb/ClickHouse:upstream_alter_database_table_override -->
This is an automated comment for commit f2428ab4876cf69b6c27ae34f2ed747d5d283540 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54219/f2428ab4876cf69b6c27ae34f2ed747d5d283540/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

kssenii commented on src/Databases/AlterDatabaseCommands.cpp:
```diff
@@ -0,0 +1,413 @@
+namespace
+{
+    template <typename T>
+    std::vector<const T *> to_const_elements(std::vector<T *> input)
```
```suggestion
    std::vector<const T *> toConstElements(std::vector<T *> input)
```

kssenii commented on src/Databases/AlterDatabaseCommands.cpp:
```diff
@@ -0,0 +1,413 @@
+    ContextMutablePtr createInternalContext(ContextPtr context)
+    {
+        ContextMutablePtr internal_context = Context::createCopy(context);
+        internal_context->setInternalQuery(true);
+        internal_context->setSetting("mutations_sync", 1);
```
Please add a comment why this is needed

stigsb replied:
I added it so that potential mutations from reconciliation queries would finish before this query returns. But it's better to let the user choose by adding `SETTINGS mutations_sync=1` to the alter database query, so I'll remove it.

kssenii commented on src/Databases/AlterDatabaseCommands.cpp:
```diff
@@ -0,0 +1,413 @@
+
+    ASTTableOverride * tryGetExistingOverride(const ASTCreateQuery * create, const String & override_name)
+    {
+        const ASTTableOverrideList * existing_table_overrides = create ? create->table_overrides : nullptr;
+        if (auto tmp = existing_table_overrides ? existing_table_overrides->tryGetTableOverride(override_name) : nullptr)
+            return tmp->as<ASTTableOverride>();
```
A bit overcomplication to my taste, isn't it simpler:
```suggestion
        if (!create || !create->table_overrides)
            return nullptr;
        if (auto table_override = create->table_overrides->tryGetTableOverride(override_name))
            return table_override->as<ASTTableOverride>();
```

kssenii commented on src/Databases/AlterDatabaseCommands.cpp:
```diff
@@ -0,0 +1,413 @@
+        return internal_context;
+    }
+
+    String whichOverrideCommand(ASTAlterCommand::Type type)
```
magic_enum::enum_name?

kssenii commented on src/Databases/AlterDatabaseCommands.cpp:
```diff
@@ -0,0 +1,413 @@
+            /// are different, the database needs to be recreated.
+            if (serializeAST(*table_override->storage) != serializeAST(*existing_override->storage))
+            {
+                apply_data.need_to_recreate_database = true;
```
why need recreate database if only one table affected? why not just recreate table?

stigsb replied:
Currently we have no way of recreating just a single table with MaterializedMySQL. I have another PR pending that will allow this.

kssenii commented on src/Databases/AlterDatabaseCommands.cpp:
```diff
@@ -0,0 +1,413 @@
+void AlterDatabaseTableOverrideCommand::apply(DatabasePtr &, ContextPtr, AlterDatabaseApplyData & apply_data)
+{
+    ASTCreateQuery * create = apply_data.getCreateQuery();
+    if (type == ASTAlterCommand::ADD_TABLE_OVERRIDE || type == ASTAlterCommand::MODIFY_TABLE_OVERRIDE)
+    {
+        if (!create->table_overrides)
```
I'd assume that if `type == ASTAlterCommand::MODIFY_TABLE_OVERRIDE` then `create->table_overrides` always `!= nullptr`, isn't it? Should this be asserted then?

stigsb replied:
Yeah, I was just being lazy. Will use chassert().

stigsb commented on src/Parsers/ASTTableOverrides.h:
```diff
@@ -24,9 +24,14 @@ class ASTTableOverride : public IAST
     ASTColumns * columns = nullptr;
     ASTStorage * storage = nullptr;
     bool is_standalone = true;
+    bool omit_empty_parens = false;
```
Parenthesis ( )

stigsb commented on src/Parsers/ASTAlterQuery.cpp:
```diff
@@ -76,10 +81,86 @@ ASTPtr ASTAlterCommand::clone() const
+{
+    switch (type)
+    {
+        case ADD_COLUMN: return "ADD_COLUMN";
```
Will fix.

stigsb committed:
Fix review comments

stigsb committed:
Implement RESET SETTING and checkAlterIsPossible for MaterializedPostgreSQL

stigsb commented:
@kssenii my PR initially didn't have a MaterializedPostgreSQL implementation, fixed that now. Took the liberty of fixing pylint warnings in .py files I touched, and added an "error_codes" helper Python module, it updates itself with new codes if you run the script directly.

stigsb committed:
black fixes

stigsb committed:
Merge branch 'master' into upstream_alter_database_table_override

stigsb committed:
AlterCommands: handle non-physical columns

stigsb committed:
Use assert_query_fails helper

stigsb committed:
fix clang-tidy warning

kssenii commented on src/Databases/PostgreSQL/DatabaseMaterializedPostgreSQL.cpp:
```diff
@@ -133,6 +134,34 @@ void DatabaseMaterializedPostgreSQL::startupTables(ThreadPool & thread_pool, Loa
+    auto check_setting_change = [&existing_settings, &query_context](const SettingChange & change)
+    {
+        const char * command = change.value.isNull() ? "RESET SETTING" : "MODIFY SETTING";
+        if (change.name == "materialized_postgresql_tables_list" && !query_context->isInternalQuery())
```
Please remove this check for `isInternalQuery`, I know it was here before and you only moved it from `applySettingsChanges`, but this check is outdated and I forgot to remove it.

kssenii commented on src/Databases/AlterDatabaseCommands.cpp:
```diff
@@ -0,0 +1,408 @@
+    ContextMutablePtr createInternalContext(ContextPtr context)
+    {
+        ContextMutablePtr internal_context = Context::createCopy(context);
+        internal_context->setInternalQuery(true);
```
Add a comment in code why this (`internal_context->setInternalQuery(true);`) is needed?

kssenii commented on src/Databases/DatabaseOnDisk.cpp:
```diff
@@ -751,8 +753,57 @@ ASTPtr DatabaseOnDisk::getCreateQueryFromStorage(const String & table_name, cons
+
+void DatabaseOnDisk::persistCreateDatabaseQueryUnlocked(const ASTPtr & ast, const DatabaseMetadataPersistedCallback & commit_cb) const
+{
+    auto database_metadata_path = getDatabaseMetadataFilePath();
```
Looks like you lost fs::canonical for `context->getPath()`, previously it was done, now it is not. Previously there was:
```
fs::path metadata_root_path = fs::canonical(query_context->getGlobalContext()->getPath());
```
and now in `getDatabaseMetadataFilePath()` there is only `getPath()`.

kssenii commented on src/Databases/DatabaseOnDisk.cpp:
```diff
@@ -751,8 +753,57 @@ ASTPtr DatabaseOnDisk::getCreateQueryFromStorage(const String & table_name, cons
+    ASTPtr ast_clone = ast->clone();
+    auto & create = ast_clone->as<ASTCreateQuery &>();
+    if (create.storage && create.storage->engine && create.storage->engine->name != getEngineName())
+    {
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Altering database engine is not allowed");
+    }
```
seems like this check should be on a different level, before we even start to do `persistCreateDatabaseQueryUnlocked`

kssenii commented on src/Databases/DatabaseOnDisk.cpp:
```diff
@@ -490,14 +490,8 @@ ASTPtr DatabaseOnDisk::getCreateDatabaseQuery() const
 
     auto settings = getContext()->getSettingsRef();
-    {
-        std::lock_guard lock(mutex);
```
why lock is no longer needed? Not sure, but looks like it was needed to avoid race on metadata in `parseQueryFromMetadata`

kssenii commented:
Hi @stigsb! Are you planning to continue with this PR or should we close it for now?

stigsb commented:
I would like to finish it, will have time July 1st

dragon-wzl commented:
Looking forward to your push!Looking forward to your completion!
